### PR TITLE
add go version to increase performance

### PR DESCRIPTION
**Pull request by Night12138**
latency jitter 50ms=>20ms

### CODE REVIEW

**Comment by tamilpp25** [comment]
I got to tell I have no experience in go but this looks really well written compared to the python one, like don't have to relaunch proxy when switching server, and the ping is low too, but I did run into a few issues while running this:

- First time login takes quite some time and KCP seems to reestablish 2-3 times and sometimes is stuck at the end of the element loading screen
- KCP doesn't close connection sometimes when quitting game (this never worked on python so I just had it save file on exiting program)
- Proxy can only save dump on connection close, which doesn't close sometimes as stated above, so possible to make it close on quitting the program?
- Packets don't seem to get processed while playing the game and client attempts to reconnect after like few seconds of entering the game

**Comment by Night12138** [comment]
@tamilpp25 I'm grateful that you take time to review it!  (maybe just have a try ) I will try to reproduce the problem at night and fix some unreasonable parts (such as strong reliance on handshake packets to close)

**Comment by tamilpp25** [comment]
Issue I saw with packets not processing was, once you enter game it works fine for like 10-20 seconds but after that it won't and ends up trying to reconnect, probably issue with kcp? as this has happened in python version multiple times like the kcp crashes midway

**Comment by Night12138** [comment]
Hi @tamilpp25, glad to sync the progress with you now. With some work I solved a bunch of issues and it's now running on my computer with amazing stability (only 2 ms latency). You can check the change with this [commit](https://github.com/MoonlightPS/Iridium-gidra/pull/1/commits/f828287e0491ad836b202c09807e7154a56c79c1).
Something missuses in my case:

1.  Didn't set `NoDelay` for the client, but I have no idea why setting `NoDelay` on your python version will cause the crash.
2. Misuse of `update()` and `check()`, using the same current time will solve the KCP always try to re-establish the connection.

BTW, now you can dump all packets when quitting server :-D.

image


Have fun!

**Comment by tamilpp25** [comment]
This is really awesome, ping is greatly reduced, and packets seem to be processed right on time but after playing for a while it looks like the kcp crashes and I get this error, this happened twice like after 2-3mins of entering game and playing around
```
Exception 0xc0000005 0x0 0x12c42abf000 0x7ff934c89749
PC=0x7ff934c89749
signal arrived during external code execution
```

also this is [where](https://github.com/MoonlightPS/Iridium-gidra/pull/1/commits/f828287e0491ad836b202c09807e7154a56c79c1#diff-c874bb81c5811aec73bd473124f225972dd11bd0f70f40a00d0f4eb35b2f12a0R81) the error seem to happen:

```
created by github.com/MoonlightPS/Iridium-gidra/gover/proxy.(*KCPConn).Start
        D:/Stuff/ps/iridium-gidra/gover/proxy/socket.go:81 +0x97
```
after which the program exits with `exit status 2`

**Comment by Night12138** [comment]
Unfortunately, this seems to be a problem with goroutine racing, maybe the KCP object is not concurrency/multi-threading safe, so I may need to lock the KCP object when reading and writing operations, which may cause refactoring... I'll watch at this issue and make some goroutine race tests tomorrow night, thanks for the feedback!

**Comment by Night12138** [comment]
Okay... KCP is not multi-threading safe, quote for https://github.com/skywind3000/kcp/issues/190#issuecomment-487355012 and https://github.com/skywind3000/kcp/issues/207, It may be the same reason for your python version crash too. I will refactor that tomorrow, now just go to bed :-P

**Comment by Night12138** [comment]
Hi @tamilpp25, I added a mutex to the operation of the KCP object in a simple way, and the game ran stably for 10+ minutes. This doesn't seem to affect performance much so I didn't refactor in a lock-free way. Please check  if the issue still exists when you are free. You may be interest to check the change with this [commit](https://github.com/MoonlightPS/Iridium-gidra/pull/1/commits/3c46cfc8a5f71db8c8773921904e0637ab281345).

**Comment by tamilpp25** [comment]
Tested for almost half an hour, the KCP doesn't crash and the gameplay is smooth 👍🏻
I might as well just replace the python version with this instead of having 2 proxies

**Comment by tamilpp25** [comment]
also I'd like you to add a few more parameters while saving the dump file so that it matches the format 
```jsonc
[
  {
    "index": int,
    "packetId": int,
    "protoName": string,
    "source": string,
    "time": float,
    "object": protobuf object
  }
]
```

**Comment by Night12138** [comment]
I will update the record format tonight, may you do more tests to find more issues I can fix together?

**Comment by tamilpp25** [comment]
sure

File: gover/utils/recorder.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+type Recorder struct {
+	q *BytesQueue
+	h *PacketHandler
+	d []*messagePack
+}
+
+func (r *Recorder) Stop() {
+	r.q.Close()
+	content, err := json.Marshal(r.d)
```
Comment (tamilpp25) [inline gover/utils/recorder.go]: replace with `content, err := json.MarshalIndent(r.d, "", "  ")` for json pretty print

Comment (Night12138) [reply gover/utils/recorder.go]: Cool, I will be replacing it with [protojson](https://pkg.go.dev/google.golang.org/protobuf/encoding/protojson), the default JSON marshaler's output is quite different from the python version, like the JSON key serializes with snake_case but not camelCase, enum using number not a string, etc...

Comment (tamilpp25) [reply gover/utils/recorder.go]: Ahh ok, I thought that was caused by protobuf go

**Comment by Night12138** [comment]
Hi @tamilpp25, I have updated the record format in this [commit](https://github.com/MoonlightPS/Iridium-gidra/pull/1/commits/dc40cbe66a7aa92e0c9f06d3459d3298eaa8edfe).

**Comment by tamilpp25** [comment]
Awesome, looks good to me and nice logger too 👍🏻

**Review by tamilpp25 — APPROVED** [review]
TODO for me:
- Use this as main and remove the python version later

**Comment by Night12138** [comment]
> TODO for me:
> 
> * Use this as main and remove the python version later

BTW, I have no experience in reversing the mhy protocol and I may be overwhelmed once mhy has updated their protocol, so maybe you may keep your python version for a while and I can keep working on translating your python code until you have got enough experience in go. :-D

**Comment by tamilpp25** [comment]
Alright sure, If you want to contact me message on discord: `tamilpp25#9288`

**Pull request merged**